File: packages/ts/src/registry.ts
/** Registry loading and name resolution. */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { NetworkError, UnknownValidatorError } from "./errors.js";

export interface Entrypoints {
  python?: string;
  ts?: string;
}

export interface RegistryEntry {
  repo: string;
  commit: string;
  entrypoints: Entrypoints;
  latency_class: "inline" | "async";
  maintainer: string;
  parity_fixture?: string;
  status: "python-only" | "ts-only" | "both";
}

export type Registry = Record<string, RegistryEntry>;

function bundledRegistryPath(): string {
  const here = path.dirname(fileURLToPath(import.meta.url));
  const dataCopy = path.join(here, "..", "data", "index.json");
  if (fs.existsSync(dataCopy)) return dataCopy;
  // packages/ts/src → repo root
  return path.resolve(here, "../../../registry/index.json");
}

function loadJsonFile(filePath: string): Registry {
  const raw = fs.readFileSync(filePath, "utf8");
  const data = JSON.parse(raw) as unknown;
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new Error(`Registry must be a JSON object: ${filePath}`);
  }
  return data as Registry;
}

async function fetchRemote(url: string): Promise<Registry> {
  try {
    if (url.startsWith("file://") || (!url.includes("://") && fs.existsSync(url))) {
      const p = url.startsWith("file://") ? fileURLToPath(url) : url;
      return loadJsonFile(p);
    }
    const resp = await fetch(url);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const data = (await resp.json()) as unknown;
    if (typeof data !== "object" || data === null || Array.isArray(data)) {
      throw new Error("remote registry is not a JSON object");
    }
    return data as Registry;
  } catch (err) {
    throw new NetworkError(`failed to fetch registry from ${url}: ${String(err)}`);
  }
}

export async function loadRegistry(opts?: {
  registryUrl?: string | null;
  bundledPath?: string;
}): Promise<Registry> {
  const bundled = loadJsonFile(opts?.bundledPath ?? bundledRegistryPath());
  const merged: Registry = { ...bundled };

  const url =
    opts && "registryUrl" in (opts ?? {})
      ? opts!.registryUrl
      : process.env.GALYA_REGISTRY_URL;

  if (url) {
    const override = await fetchRemote(url);
    Object.assign(merged, override);
  }
  return merged;
}

export function resolve(name: string, registry: Registry): RegistryEntry {
  const entry = registry[name];
  if (!entry) throw new UnknownValidatorError(name);
  return entry;
}

export function availableLanguages(entry: RegistryEntry): string[] {
  const eps = entry.entrypoints ?? {};
  return (["python", "ts"] as const).filter((lang) => Boolean(eps[lang]));
}

File: packages/python/src/galya_reality/registry.py
"""Registry loading and name resolution."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, TypedDict

import httpx

from galya_reality.exceptions import NetworkError, UnknownValidatorError


class Entrypoints(TypedDict, total=False):
    python: str
    ts: str


class RegistryEntry(TypedDict, total=False):
    repo: str
    commit: str
    entrypoints: Entrypoints
    latency_class: str
    maintainer: str
    parity_fixture: str
    status: str


def _bundled_registry_path() -> Path:
    """Locate the shipped registry JSON.

    Prefer a package-data copy (installed wheel), then fall back to the
    monorepo ``registry/index.json`` for editable / source checkouts.
    """
    data = Path(__file__).resolve().parent / "data" / "index.json"
    if data.is_file():
        return data
    # packages/python/src/galya_reality → repo root (4 levels up)
    repo_root = Path(__file__).resolve().parents[4]
    return repo_root / "registry" / "index.json"


def _load_json_file(path: Path) -> dict[str, Any]:
    with path.open(encoding="utf-8") as fh:
        data = json.load(fh)
    if not isinstance(data, dict):
        raise ValueError(f"Registry must be a JSON object, got {type(data)}")
    return data


def _fetch_remote(url: str) -> dict[str, Any]:
    try:
        resp = httpx.get(url, timeout=30.0, follow_redirects=True)
        resp.raise_for_status()
        data = resp.json()
    except Exception as exc:  # noqa: BLE001 — surface as NetworkError
        raise NetworkError(f"failed to fetch registry from {url}: {exc}") from exc
    if not isinstance(data, dict):
        raise NetworkError(f"remote registry at {url} is not a JSON object")
    return data


def load_registry(
    *,
    registry_url: str | None = None,
    bundled_path: Path | None = None,
) -> dict[str, RegistryEntry]:
    """Load and merge the bundled registry with an optional URL override.

    Entries from ``GALYA_REGISTRY_URL`` (or ``registry_url``) take
    precedence over the bundled index for the same keys.
    """
    path = bundled_path or _bundled_registry_path()
    merged: dict[str, RegistryEntry] = dict(_load_json_file(path))  # type: ignore[arg-type]

    url = registry_url if registry_url is not None else os.environ.get("GALYA_REGISTRY_URL")
    if url:
        override = _fetch_remote(url)
        for key, value in override.items():
            if isinstance(value, dict):
                merged[key] = value  # type: ignore[assignment]
    return merged


def resolve(name: str, registry: dict[str, RegistryEntry] | None = None) -> RegistryEntry:
    """Look up a validator by name. Raises ``UnknownValidatorError`` if missing."""
    reg = registry if registry is not None else load_registry()
    if name not in reg:
        raise UnknownValidatorError(name)
    return reg[name]


def available_languages(entry: RegistryEntry) -> list[str]:
    """Return the language keys present under ``entrypoints``."""
    eps = entry.get("entrypoints") or {}
    return [lang for lang in ("python", "ts") if lang in eps and eps[lang]]
